flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/OperatorEventValve.java

```diff
@@ -96,42 +94,44 @@ public boolean isShut() {
      * (because it gets dropped through a call to {@link #reset()} or {@link #resetForTask(int)},
      * then the returned future till be completed exceptionally.
      */
-    public CompletableFuture<Acknowledge> sendEvent(
-            SerializedValue<OperatorEvent> event, int subtask) {
-        synchronized (lock) {
-            if (!shut) {
-                return eventSender.apply(event, subtask);
-            }
-
-            final List<BlockedEvent> eventsForTask =
-                    blockedEvents.computeIfAbsent(subtask, (key) -> new ArrayList<>());
-            final CompletableFuture<Acknowledge> future = new CompletableFuture<>();
-            eventsForTask.add(new BlockedEvent(event, subtask, future));
-            return future;
+    public void sendEvent(
+            SerializedValue<OperatorEvent> event,
+            int subtask,
+            CompletableFuture<Acknowledge> result) {
+        checkRunsInMainThread();
+
+        if (!shut) {
+            final CompletableFuture<Acknowledge> ack = eventSender.apply(event, subtask);
```

Comment:
Should we also add `assertRunningInJobMasterMainThread()` to `Execution.sendOperatorEvent` in order to guard that this method is only accessed from the main thread?

Reply:
I guess not. `Execution.sendOperatorEvent` is capable to run concurrently and the locking version(eg. revision before this change) should also work. It is just a matter of `OperatorCoordinator`'s choice.

Reply:
I am actually neutral to either thread modeling in `OperatorCoordinator`, it is pure implementation detail. No matter how we choose, rendezvous between checkpoint completion and event sending is unavoidable.

Reply:
I think we don't need the guard in `Execution.sendOperatorEvent` because at that point it doesn't matter. The RPC gateways are thread safe. We only need to make sure we don't reach the wrong target, and for that we only need to reference the right `Execution` object (that is done in the next PR).

Now that I think about it a bit more. We also don't really need the scheduler main thread executor for sending, we really only need to know that all event sending and checkpoint completion actions are done in the same thread so that order is preserved. The fact that this thread is the scheduler main-thread is an implementation detail that was chosen to make other parts simpler.

Reply:
tl;dr - I would skip the check there, it enforces a contract we don't need and might make future changes more difficult, when trying to keep guaranteeing an unneeded contract.

Reply:
I am not entirely sure about this. The problem in my opinion is that `Execution` is not thread-safe. `sendOperatorEvent` accesses internal state of the `Execution` which is set by another thread (`assignedResource` and `getState()`). Since there is no synchronization it is not guaranteed that all changes are visible.

Moreover, it breaks a bit with the existing threading model of the `ExecutionGraph` where we said that there is only a single thread accessing it.

Reply:
Fair enough, then let's add an extra check there as well. If it is a useful check regardless of the contract here.

Reply:
I'll add this as part of #15601 (where I have to incorporate comments anyways) to avoid re-running this PR again on CI (my CI queue backlog is long already...).

Reply:
Skimming through `sendOperatorEvent`, I think it is thread safe for `SubtaskGateway`. The points are:
* After `subtaskReady` and `RUNNING`(or `INITIALIZING`), `assignedResource` is visible anyway. Thus, not sending will lose.
* Up failing, which kind of failure actually does not matter.

Though, this "thread-safe" could be an accident.

Reply:
Yes, I think it will work in practice. But @tillrohrmann is right, it is unexpected, it goes against how things are done in the scheduler. So let's catch that situation, because there is a chance that when this comes from a different thread, it is not intentional, and the check can help us debug this faster.